docs/audit_report/src/02_security_issues.rst

```diff
@@ -28,3 +28,82 @@ exploitable by an attacker under certain circumstances.
+encoded OID (of a named curve) or encoded parameters of a custom curve. The
+disambiguation is performed internally. This could allow an attacker to
+masquerade a custom curve as an OID of a named curve and exploit the
+vulnerability.
+
+Note that these constructors are called also internally and may receive
+user-provided or potentially untrusted data. For instance, when loading a
+maliciously crafted key in an ASN.1 container or even embedded into an X.509
+certificate.
+
+For a vulnerable server requiring TLS client authentication this
+might be a feasible denial of service amplifier without requiring the attacker
+to be in a privileged network location. An attacker on the network path, could
+mount a similar attack by impersonating the server and sending a malicious
+certificate to waste CPU time on the client and cause a denial of service.
+
+Some of the interfaces that internally call the above constructors are:
+
+* ``create_private_key(std::string_view alg_name, RandomNumberGenerator& rng,
+  std::string_view params, std::string_view provider)`` where the `params`
+  argument may contain custom curve parameters.
+* ``EC_PrivateKey(const AlgorithmIdentifier& alg_id, std::span<uint8_t>
+  key_bits, bool with_modular_inverse)`` where the `alg_id` argument may contain
+  encoded custom curve parameters in its `parameters` field.
+* ``EC_PublicKey(const AlgorithmIdentifier& alg_id, std::span<uint8_t>
+  key_bits)`` where the `alg_id` argument may contain encoded custom curve
+  parameters in its `parameters` field.
+* ``X509::load_key(...)`` (all of its overloads), with a custom curve encoded
+  into a key container format. This is used to load public keys embedded in
+  X.509 certificates.
+
+This issue was reported to the maintainer by Bing Shi.
+
+.. _secinfo/binary_gcd:
+
+Faulty gcd() implementation
+---------------------------
+
+As shown by the OSS-Fuzz project, Botan's previous implementation to find the
+greatest common denominator of two arbitrarily sized integers generated an error
+for certain inputs. The GCD algorithm was based on "safegcd" [SAFEGCD]_, but
+with a slight adaption to avoid the requirement of constant-time arithmetics on
+signed integers. This adaption violated the algorithm's termination assumptions,
+causing the sporadic error.
+
+Most notably, the gcd() function is used in the RSA key generation. As a result,
+key generation might have sporadically failed with an exception. In the worst
+case this could have caused a bias in the RSA key pairs generated by Botan.
+However, we are not aware of any failures in a production environment.
+
+As a permanent fix, Botan |botan_version| uses the binary GCD algorithm
+[BINARYGCD]_, which does not require constant-time arithmetics on signed intgers
+and provides a termination assumption that is much easier to reason about.
+Additionally, the gcd() now leaks less information via timing side channels as
+before.
```

Comment:
This sounds a bit like we fixed some of the many side channels. Maybe we want to write something like:
```suggestion
Additionally, the new gcd() is more resistant against timing side channels than the previous version.
```